PublicMapping/district-builder-2
# Backend skeleton

## Overview

Sets up a backend skeleton project in NestJS. Also ensures code-sharing works between the frontend and backend, which required some reorganization. A database container has also been setup, along with a User model and CRUD route, and demonstration of it's usage in the frontend.

### Demo

![screenshot_2](https://user-images.githubusercontent.com/6386/73463163-cb8edd80-434a-11ea-9bdb-b7f4bc30f123.png)

### Notes

For code-sharing, there are two things being shared and used across the client and server:
 1) An interface for the User model
 2) A test function which returns the "It worked!" string. I left this in here for now, because I saw some false-positives when testing code sharing when using interfaces-only (i.e. CRA wasn't complaining in scenarios where sharing functions was failing, due to how it integrates with `tsc`). I think we should leave it in until we have a real shared function.

## Testing Instructions
 - Run `scripts/update`, which should build Docker containers and run migrations
 - In one terminal, run: `scripts/server`
 - In another terminal, run: `scripts/client`
 - Browse to `http://localhost:3003/`. Should see the `It worked!` message, and no users.
 - In another terminal, add a user via the CRUD endpoint: 
```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"email": "testuser@example.com"}' \
  http://localhost:3005/users
```
- Refresh the page and you should see your user listed (similar to the screenshot) 

Closes #3
Closes #23 
Closes #24
